### PR TITLE
BUG: bug in GroupBy.count where arg minlength passed to np.bincount must be None for np<1.13

### LINKED ISSUE #21956: BUG: GroupBy.count arg minlength passed to np.bincount must be None rather than 0 for np<1.13
```python
>>> df = pd.DataFrame({'A': [np.nan, np.nan], 'B': ['a', 'b'], 'C':[1,2]})
>>> df.groupby(['A', 'B']).C.count()
Series([], Name: C, dtype: int64)  # ok with numpy>=1.13
ValueError: minlength must be positive  # with numpy <1.13
```

#### Problem description

Currently, ``SeriesGroupBy.count`` passes 0 to np.bincount. This is ok for numpy>=1.13, where 0 is an accepted value. However, for numpy<1.13, 0 was not a possible value for minlength, and None should be passed instead. 

This is easy to fix, in ``SeriesGroupBy.count`` just insert a guard for numpy<1.13 and minlength=0.

#### Expected Output

``Series([], Name: C, dtype: int64)``, irrelevant of numpy version.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 664ddb84316a6e683b0338e4bd5e8279530b6b73
python: 3.6.3.final.0
python-bits: 32
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.24.0.dev0+319.g2f2876b8c.dirty
pytest: 3.3.1
pip: 10.0.1
setuptools: 38.2.5
Cython: 0.28.3
numpy: 1.13.3
scipy: 1.0.0
pyarrow: None
xarray: 0.10.0
IPython: 6.4.0
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.0
openpyxl: 2.4.9
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: None
bs4: None
html5lib: 1.0b10
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #21956
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

See #21956 for details.

### CODE REVIEW

File: pandas/tests/groupby/test_counting.py [hunk 1]
```diff
@@ -212,3 +212,13 @@ def test_count_with_datetimelike(self, datetimelike):
         expected = DataFrame({'y': [2, 1]}, index=['a', 'b'])
         expected.index.name = "x"
         assert_frame_equal(expected, res)
+
+    def test_count_with_only_nans_in_first_group(self):
+        # GH21956
+        df = DataFrame({'A': [np.nan, np.nan], 'B': ['a', 'b'], 'C': [1, 2]})
+        result = df.groupby(['A', 'B']).C.count()
```
Comment (jreback) [inline pandas/tests/groupby/test_counting.py]: can u do an assert_produces_warning()

here (should not show a warning)

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1207,7 +1208,10 @@ def count(self):
 
         mask = (ids != -1) & ~isna(val)
         ids = ensure_platform_int(ids)
-        out = np.bincount(ids[mask], minlength=ngroups or 0)
+        minlength = ngroups or 0
```
Comment (jreback) [inline pandas/core/groupby/generic.py]: can u simplify this

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_counting.py (hunk 1 above)
Comment (topper-123) [reply pandas/tests/groupby/test_counting.py]: This does not produce any warnings for me.

What should look for in travis? Grep "SeriesGroupBy.count"?

Comment (topper-123) [reply pandas/tests/groupby/test_counting.py]: Ok, will look in the morning.

File: pandas/core/groupby/generic.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/groupby/generic.py]: I've made this a one-liner in the newest update.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.4.txt [hunk 3]
```diff
@@ -31,6 +31,7 @@ Bug Fixes
 **Groupby/Resample/Rolling**
 
 - Bug where calling :func:`DataFrameGroupBy.agg` with a list of functions including ``ohlc`` as the non-initial element would raise a ``ValueError`` (:issue:`21716`)
+- Bug in :func:`pandas.core.groupby.SeriesGroupBy.count` when using numpy < 1.13 and ngroups=0 (:issue:`21956`).
```
[the remark this answers is not in the document]
Comment (topper-123) [reply doc/source/whatsnew/v0.23.4.txt]: Yes, as the per the example in #21956:

```python
>>> df = pd.DataFrame({'A': [np.nan, np.nan], 'B': ['a', 'b'], 'C':[1,2]})
>>> df.groupby(['A', 'B']).C.count()
ValueError: minlength must be positive  # when numpy <1.13
```

File: pandas/core/groupby/generic.py [hunk 4]
```diff
@@ -1207,7 +1208,8 @@ def count(self):
 
         mask = (ids != -1) & ~isna(val)
         ids = ensure_platform_int(ids)
-        out = np.bincount(ids[mask], minlength=ngroups or 0)
+        minlength = ngroups or (None if _np_version_under1p13 else 0)
```
[the remark this answers is not in the document]
Comment (topper-123) [reply pandas/core/groupby/generic.py]: I think you are right, th doc string says "minlength : int, optional" and it accepts None for numpy >1.13. I'll change it.

File: doc/source/whatsnew/v0.23.4.txt [hunk 5]
```diff
@@ -32,6 +32,7 @@ Bug Fixes
 
 - Bug where calling :func:`DataFrameGroupBy.agg` with a list of functions including ``ohlc`` as the non-initial element would raise a ``ValueError`` (:issue:`21716`)
 - Bug in ``roll_quantile`` caused a memory leak when calling ``.rolling(...).quantile(q)`` with ``q`` in (0,1) (:issue:`21965`)
+- Bug in :func:`pandas.core.groupby.SeriesGroupBy.count` when using numpy < 1.13 and ngroups=0 (:issue:`21956`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.4.txt]: again, is this actually a user facing change? (not against the note), just this is not understandable to user.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.4.txt (hunk 5 above)
Comment (topper-123) [reply doc/source/whatsnew/v0.23.4.txt]: Ok, new explanation uploaded.

File: pandas/tests/groupby/test_counting.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/groupby/test_counting.py]: ok, can you assert that this produces no warning (e.g. it IS producing a warning now in numpy < 1.13)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by topper-123** [comment]
@jreback I got a straight up ValueError in master, so no warning. That ValueError is not in this PR, obviously

I've got a feeling I don't understand you. Could you give an example where a warning is emitted?

File: doc/source/whatsnew/v0.23.4.txt [hunk 6]
```diff
@@ -32,6 +32,8 @@ Bug Fixes
 
 - Bug where calling :func:`DataFrameGroupBy.agg` with a list of functions including ``ohlc`` as the non-initial element would raise a ``ValueError`` (:issue:`21716`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.4.txt]: move this to 0.24.0, as this wont' backport cleanly.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]
lgtm. ping on green.

**Comment by topper-123** [comment]
Something's wrong with travis, else green:

```
Worker information
hostname: 1773d14a-144b-4458-b6fe-9060bb57620b@1.production-2-worker-org-09-packet
version: v4.0.0 https://github.com/travis-ci/worker/tree/e5cb567e10c0eefe380e81c9a2229ac8fb6a16ce
instance: 2a708b5 travisci/ci-garnet:packer-1512502276-986baf0 (via amqp)
startup: 2.24796375s
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

**Comment by jreback** [comment]
thanks @topper-123

**Comment by jreback** [comment]
so these are the warnings:

```
/home/travis/build/pandas-dev/pandas/pandas/core/groupby/generic.py:1209: DeprecationWarning: 0 should be passed as minlength instead of None; this will error in future.
    out = np.bincount(ids[mask], minlength=ngroups or None)
```

can you PR to fix this. I guess we have to switch on the version and actually pass 0 rather than None for > 113

**Pull request merged**